Supporting SMART App Launch with Keycloak and the IBM FHIR Server (#35)
Alvearie/alvearie.github.io: src/pages/blog/smart-keycloak/index.mdx

This specification extends OAuth 2.0 with custom `scopes` and the notion of `launch context`.

SMART *scopes* are normal OAuth 2.0 scope strings that are used to define an authorization model that is aligned with FHIR resources.
The specification defines the scope strings via a pattern like `<context>/<resourceType>.<permission>`, where
- *context* is `patient` or `user`;
- *resourceType* is one of the resource types defined in the HL7 FHIR specification, or `*` to indicate all resource types; and
- *permission* is `read` or `write`, or `*` for read and write.

SMART *launch context* is what allows a user to launch an application in the context of a given user session or patient record.
SMART App Launch defines two different launch varieties: one for launching apps from an Electronic Health Record (EHR) system and one for launching "standalone" apps.

In this post, I walk through how to support standalone app launch with the IBM FHIR Server via [Keycloak](https://www.keycloak.org), 
an open source identity and access management solution.

### Introducing the Alvearie Keycloak Extensions for FHIR

> Keycloak is wonderfully-extensible and easily-configurable.

Reviewer: remove the - marks here

should be "wonderfully extensible and easily configurable"

> Keycloak is wonderfully-extensible and easily-configurable.

Reviewer: ```suggestion
Keycloak is wonderfully extensible and easily configurable.
```

### Using the Alvearie Keycloak Extensions for FHIR
The Keycloak extensions for FHIR project currently consists of two components:
- *keycloak-config* for automating the configuration of a Keycloak realm in support of standalone app launch with patient context
- *keycloak-extensions* which defines a set of extensions that fill the gaps in Keycloak's support of SMART App Launch:
  - an AudienceValidator authenticator for validating that the `aud` parameter passed to the server matches the desired value
  - a UserAttributeMapper mapper for mapping user attributes into custom fields in the token response payload (rather than claims in the issued tokens)

To use the extensions:
1. Build the keycloak-extensions project into a jar file and add it to the [`standalone/deployments`](https://www.keycloak.org/docs/latest/server_development/#packaging-classes-and-deployment-2) directory of your Keycloak installation.

>     - Note: moving forward, we hope to add continuous integration support and publish the built artifacts to Maven Central.

Author: ```suggestion
    - Note: moving forward, we hope to publish the built artifacts to Maven Central.
```

            "patient/ExplanationOfBenefit.read",
            "patient/Goal.read",
            "patient/Immunization.read",
            "patient/Location.read",
            "patient/Medication.read",
            "patient/MedicationRequest.read",
            "patient/Observation.read",
            "patient/Organization.read",
            "patient/Patient.read",
            "patient/Practitioner.read",
            "patient/PractitionerRole.read",
            "patient/Procedure.read",
            "patient/Provenance.read",
            "patient/RelatedPerson.read"
          ],
          "capabilities": [
            "sso-openid-connect",
            "launch-standalone",
            "client-public",
            "client-confidential-symmetric",
            "permission-offline",
            "context-standalone-patient",
            "permission-patient"
          ]
        }
      }
    },
...
```

### Conclusion 
In this post, we walked through the OAuth 2.0 extensions defined by SMART App Launch, how to support those extensions via the Alvearie Keycloak Extensions for FHIR, and how to use this project together with the IBM FHIR Server to support SMART App Launch.

> SMART App Launch is just one of the specification defined by SMART-on-FHIR and the authentication and authorization challenges in healthcare are immense. We'd love to hear from members of the FHIR community, the SMART-on-FHIR community, and the Keycloak community on the approach outlined here and where you'd like to see us go next. Please connect with us on [chat.fhir.org](https://chat.fhir.org/#narrow/stream/179170-smart/topic/Keycloak.20for.20SMART.20authz), the [Keycloak Extensions for FHIR project](https://github.com/Alvearie/keycloak-extensions-for-fhir), or the [IBM FHIR Server project](https://github.com/ibm/fhir) about how we can take this to the next level.

Reviewer: change to plural form - "specifications"

> SMART App Launch is just one of the specification defined by SMART-on-FHIR and the authentication and authorization challenges in healthcare are immense. We'd love to hear from members of the FHIR community, the SMART-on-FHIR community, and the Keycloak community on the approach outlined here and where you'd like to see us go next. Please connect with us on [chat.fhir.org](https://chat.fhir.org/#narrow/stream/179170-smart/topic/Keycloak.20for.20SMART.20authz), the [Keycloak Extensions for FHIR project](https://github.com/Alvearie/keycloak-extensions-for-fhir), or the [IBM FHIR Server project](https://github.com/ibm/fhir) about how we can take this to the next level.

Reviewer: ```suggestion
SMART App Launch is just one of the specifications defined by SMART-on-FHIR and the authentication and authorization challenges in healthcare are immense. We'd love to hear from members of the FHIR community, the SMART-on-FHIR community, and the Keycloak community on the approach outlined here and where you'd like to see us go next. Please connect with us on [chat.fhir.org](https://chat.fhir.org/#narrow/stream/179170-smart/topic/Keycloak.20for.20SMART.20authz), the [Keycloak Extensions for FHIR project](https://github.com/Alvearie/keycloak-extensions-for-fhir), or the [IBM FHIR Server project](https://github.com/ibm/fhir) about how we can take this to the next level.
```

---

> SMART is the registered trademarks of SMART Technologies ULC in the U.S. and/or other countries.

Reviewer 2: Perhaps it should follow https://smarthealthit.org/smart-logos-guidelines-for-trademark-use/
I don't think the above is accurate.

Author: Good catch.  I just googled SMART trademark and followed https://www.smarttech.com/en/legal/trademark-guidelines...didn't even realize that there are two entirely different groups claiming trademark on "SMART"...seems problematic.

Author: ```suggestion
SMART is a trademark of The Children’s Medical Center Corporation.
```
